cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -1203,6 +1203,47 @@ inline Status ConvertFixedSizeBinary(const ChunkedArray& data, PyObject** out_va
   return Status::OK();
 }
 
+inline Status ConvertStruct(const ChunkedArray& data, PyObject** out_values) {
+  PyAcquireGIL lock;
+  for (int c = 0; c < data.num_chunks(); c++) {
+    auto arr = static_cast<StructArray*>(data.chunk(c).get());
+    std::vector<PyObject*> fields_data;
+    // Convert the struct arrays first
+    for (auto& field : arr->fields()) {
+      PyObject* numpy_array;
+      RETURN_NOT_OK(ConvertArrayToPandas(field, nullptr, &numpy_array));
+      fields_data.push_back(numpy_array);
+    }
+
+    // Construct a dictionary for each row
+    const bool has_nulls = data.null_count() > 0;
+    for (int64_t i = 0; i < arr->length(); ++i) {
+      if (has_nulls && arr->IsNull(i)) {
+        Py_INCREF(Py_None);
+        *out_values = Py_None;
+      } else {
+        PyObject* item = PyDict_New();
+        for (uint field_idx = 0; field_idx < arr->fields().size(); field_idx++) {
+          if (!arr->field(field_idx)->IsNull(i)) {
+            auto name = arr->type()->child(field_idx)->name();
+            PyObject* field_value = PyObject_GetItem(fields_data[field_idx],
+              PyLong_FromLong(i));
```

Comment:
`PyLong_FromLong` returns a new reference, so this will leak memory. You can use [`OwnedRef`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/common.h#L66) to automatically call `Py_XDECREF` on your object on scope exit.